tikv/region.go

```diff
@@ -176,10 +177,147 @@ type RegionOptions struct {
 	RegionSize int64
 }
 
-type RegionManager struct {
-	storeMeta  metapb.Store
-	mu         sync.RWMutex
-	regions    map[uint64]*regionCtx
+type RegionManager interface {
+	GetRegionFromCtx(ctx *kvrpcpb.Context) (*regionCtx, *errorpb.Error)
+	CheckLeader(ctx *kvrpcpb.Context) *errorpb.Error
+	Close() error
+}
+
+type regionManager struct {
+	storeMeta *metapb.Store
+	mu        sync.RWMutex
+	regions   map[uint64]*regionCtx
+}
+
+func (rm *regionManager) checkContext(ctx *kvrpcpb.Context) *errorpb.Error {
+	ctxPeer := ctx.GetPeer()
+	if ctxPeer != nil && ctxPeer.GetStoreId() != rm.storeMeta.Id {
+		return &errorpb.Error{
+			Message:       "store not match",
+			StoreNotMatch: &errorpb.StoreNotMatch{},
+		}
+	}
+	return nil
+}
+
+func (rm *regionManager) GetRegionFromCtx(ctx *kvrpcpb.Context) (*regionCtx, *errorpb.Error) {
+	if err := rm.checkContext(ctx); err != nil {
+		return nil, err
+	}
+	rm.mu.RLock()
+	ri := rm.regions[ctx.RegionId]
+	if ri != nil {
+		ri.refCount.Add(1)
+	}
+	rm.mu.RUnlock()
+	if ri == nil {
+		return nil, &errorpb.Error{
+			Message: "region not found",
+			RegionNotFound: &errorpb.RegionNotFound{
+				RegionId: ctx.GetRegionId(),
+			},
+		}
+	}
+	// Region epoch does not match.
+	if rm.isEpochStale(ri.meta.GetRegionEpoch(), ctx.GetRegionEpoch()) {
+		ri.refCount.Done()
+		return nil, &errorpb.Error{
+			Message: "stale epoch",
+			EpochNotMatch: &errorpb.EpochNotMatch{
+				CurrentRegions: []*metapb.Region{ri.meta},
+			},
+		}
+	}
+	ri.waitParent()
+	return ri, nil
+}
+
+func (rm *regionManager) isEpochStale(lhs, rhs *metapb.RegionEpoch) bool {
+	return lhs.GetConfVer() != rhs.GetConfVer() || lhs.GetVersion() != rhs.GetVersion()
+}
+
+type RaftRegionManager struct {
+	regionManager
+	router   raftstore.RaftstoreRouter
+	checkers map[uint64]raftstore.LeaderChecker
```

Comment:
If we put `raftstore.LeaderChecker` into `regionCtx`, then we don't need to access two maps under lock.